File: market-application/src/main/java/com/example/market/application/service/MarkPayoutCompletedService.java
package com.example.market.application.service;

import com.example.market.application.exception.PayoutNotFoundException;
import com.example.market.application.port.in.MarkPayoutCompletedUseCase;
import com.example.market.application.port.out.EventPublisher;
import com.example.market.application.port.out.PayoutRepository;
import com.example.market.domain.settlement.Payout;
import com.example.market.domain.settlement.PayoutId;
import com.example.market.domain.settlement.PayoutStatus;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;

@Service
@RequiredArgsConstructor
@Slf4j
public class MarkPayoutCompletedService implements MarkPayoutCompletedUseCase {

    private final PayoutRepository payouts;
    private final EventPublisher events;
    private final Clock clock;

    @Override
    @Transactional
    public void markCompleted(PayoutId payoutId) {
        Payout payout = payouts.findById(payoutId)
                .orElseThrow(() -> new PayoutNotFoundException(payoutId));
        if (payout.status() == PayoutStatus.COMPLETED) {
            log.info("payout already completed — idempotent skip {}", payoutId);
            return;
        }
        var ev = payout.complete(clock.instant());
        payouts.save(payout);
        events.publish(ev);
        log.info("payout completed {}", payoutId);
    }
}

File: market-application/src/main/kotlin/com/example/market/application/service/MarkPayoutCompletedService.kt
package com.example.market.application.service

import com.example.market.application.exception.PayoutNotFoundException
import com.example.market.application.port.`in`.MarkPayoutCompletedUseCase
import com.example.market.application.port.out.EventPublisher
import com.example.market.application.port.out.PayoutRepository
import com.example.market.domain.settlement.PayoutId
import com.example.market.domain.settlement.PayoutStatus
import java.time.Clock
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class MarkPayoutCompletedService(
    private val payouts: PayoutRepository,
    private val events: EventPublisher,
    private val clock: Clock,
) : MarkPayoutCompletedUseCase {

    @Transactional
    override fun markCompleted(payoutId: PayoutId) {
        val payout = payouts.findById(payoutId)
            .orElseThrow { PayoutNotFoundException(payoutId) }
        if (payout.status == PayoutStatus.COMPLETED) {
            log.info("payout already completed — idempotent skip {}", payoutId)
            return
        }
        val ev = payout.complete(clock.instant())
        payouts.save(payout)
        events.publish(ev)
        log.info("payout completed {}", payoutId)
    }

    companion object {
        private val log = LoggerFactory.getLogger(MarkPayoutCompletedService::class.java)
    }
}
